ArjunaJTA/jta/tests/classes/com/hp/mwtests/ts/jta/jca/XATerminatorUnitTest.java

```diff
@@ -417,6 +423,142 @@ public void testConcurrentImport () throws Exception {
         assertEquals("some transaction import futures did not complete", completionCount.get(), TASK_COUNT);
     }
 
+    @Test
+    public void testRecovery() throws Exception {
+        Implementations.initialise();
+        XidImple xid = new XidImple(new Uid());
+        TransactionImporter imp = SubordinationManager.getTransactionImporter();
+
+        SubordinateTransaction importTransaction = imp.importTransaction(xid);
+
+        XATerminatorImple xa = new XATerminatorImple();
+        importTransaction.enlistResource(new XAResource() {
+
+            @Override
+            public void start(Xid xid, int flags) throws XAException {
+            }
+
+            @Override
+            public void end(Xid xid, int flags) throws XAException {
+            }
+
+            @Override
+            public int prepare(Xid xid) throws XAException {
+                return 0;
+            }
+
+            @Override
+            public void commit(Xid xid, boolean onePhase) throws XAException {
+                throw new XAException(XAException.XA_RETRY);
+            }
+
+            @Override
+            public void rollback(Xid xid) throws XAException {
+            }
+
+            @Override
+            public void forget(Xid xid) throws XAException {
+            }
+
+            @Override
+            public Xid[] recover(int flag) throws XAException {
+                return null;
+            }
+
+            @Override
+            public boolean isSameRM(XAResource xaRes) throws XAException {
+                return false;
+            }
+
+            @Override
+            public int getTransactionTimeout() throws XAException {
+                return 0;
+            }
+
+            @Override
+            public boolean setTransactionTimeout(int seconds) throws XAException {
+                return false;
+            }
+        });
+
+        assertTrue(xa.beforeCompletion(xid));
+
+        assertEquals(xa.prepare(xid), XAResource.XA_OK);
+
+        try {
+            xa.commit(xid, false);
+            fail();
+        } catch (XAException e) {
+            assertTrue(e.errorCode == XAException.XAER_RMFAIL);
+        }
+
+        Xid[] recover = xa.recover(XAResource.TMSTARTRSCAN);
```

Comment:
Now that you are using the result (to assert that there is something to recover) the call makes sense.